### PR TITLE
xds: stop expiry timer after receiving an update

### PR DESCRIPTION
**Pull request by menghanl**

### CODE REVIEW

File: xds/internal/client/client_watchers_cluster_test.go [hunk 1]
```diff
@@ -274,3 +274,45 @@ func (s) TestClusterWatchExpiryTimer(t *testing.T) {
...
+	c.WatchCluster(testCDSName, func(u ClusterUpdate, err error) {
+		clusterUpdateCh.Send(clusterUpdateErr{u: u, err: err})
+	})
+
+	wantUpdate := ClusterUpdate{ServiceName: testEDSName}
+	v2Client.r.newCDSUpdate(map[string]ClusterUpdate{
+		testCDSName: wantUpdate,
+	})
+
+	if u, err := clusterUpdateCh.Receive(); err != nil || u != (clusterUpdateErr{wantUpdate, nil}) {
```
Comment (easwars) [inline xds/internal/client/client_watchers_cluster_test.go:309]: Do you really those parentheses around `clusterUpdateErr{}`?

File: xds/internal/client/client_watchers_cluster_test.go [hunk 2]
```diff
@@ -274,3 +274,45 @@ func (s) TestClusterWatchExpiryTimer(t *testing.T) {
...
+	v2ClientCh, cleanup := overrideNewXDSV2Client()
+	defer cleanup()
+
+	c, err := New(clientOpts(testXDSServer))
+	if err != nil {
+		t.Fatalf("failed to create client: %v", err)
+	}
+	defer c.Close()
+
+	v2Client := <-v2ClientCh
```
Comment (easwars) [inline xds/internal/client/client_watchers_cluster_test.go:297]: Should we add a timer around this receive as well, so we fail early if new v2client is not created?

File: xds/internal/client/client_callback.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ func (c *Client) scheduleCallback(wi *watchInfo, update interface{}, err error)
 
 func (c *Client) callCallback(wiu *watcherInfoWithUpdate) {
 	c.mu.Lock()
+	wiu.wi.expiryTimer.Stop()
```
Comment (easwars) [inline xds/internal/client/client_callback.go:37]: Should we put this in `scheduleCallback` instead?

File: xds/internal/client/client_watchers_cluster_test.go (hunk 2 above)
Comment (menghanl) [reply xds/internal/client/client_watchers_cluster_test.go:297]: I added a TODO. This will be a cleanup for all tests.
And this won't be necessary if we resolve another `TODO` to not call `v2Client.r.newCDSUpdate`, and instead do `c.newCDSUpdate`.

File: xds/internal/client/client_callback.go (hunk 3 above)
Comment (easwars) [reply xds/internal/client/client_callback.go:37]: This is the sequence of events I'm thinking about
- watcher registered at time = 0
- response received at `expiryTime - delta`
- we schedule a callback with the results from the response
- `delta` passed by
- the expiry timer is stopped and another callback is scheduled with an error saying watcher timeout

Eventually, application will receive both callbacks.

Comment (menghanl) [reply xds/internal/client/client_callback.go]: That's right! And I also found that expiry timer doesn't need the mutex. I moved it to `scheduleCallback()`.

This reduces the window, but race can still happen when timeout races with the response. Because timer.Stop() doesn't ensure the callback won't happen (even with `if !timer.Stop { <- timer.C }`)

To really solve this problem, we need a boolean, and skip timeout error. I added a commit, PTAL.

Comment (menghanl) [reply xds/internal/client/client_callback.go:41]: Hmm, I moved `timer.Stop` back to `callCallback()`, because race detector was complaining.

The race is between the afterFunc body, and assigning to expiryTimer. It could happen if timeout is very short, so func (which read timer variable) is executed at the same time the timer variable is assigned.
```go
	wi.expiryTimer = time.AfterFunc(defaultWatchExpiryTimeout, func() {
		c.scheduleCallback(wi, ClusterUpdate{}, errCallbackTimeout)
	})
```

File: xds/internal/client/client_callback.go [hunk 4]
```diff
@@ -18,22 +18,46 @@
...
+		wi.expiryTimer.Stop()
+	}
 	c.updateCh.Put(&watcherInfoWithUpdate{
 		wi:     wi,
 		update: update,
 		err:    err,
 	})
 }
 
+var errCallbackTimeout = fmt.Errorf("callback timeout")
```
Comment (easwars) [inline xds/internal/client/client_callback.go:42]: Prefer using `errors.New` since there are no formatting directives.

File: xds/internal/client/client_callback.go (hunk 3 above)
Comment (easwars) [reply xds/internal/client/client_callback.go]: This is actually quite hard to follow now with things happening at multiple places. How about doing the following:
- add a `state` enum to the `watchInfo` struct with possible values being `{created, response_received, timer_expired, cancelled}`. And this `state` will be protected by a mutex.
- in the `AfterFunc`, you grab the mutex and check the `state` and if it is already in `response_received`, you do nothing. You dont even queue a callback.
- in the `newxDSUpdate` funcs, before queuing the callback, we set `state` to `response_received`, stop the timer and queue the callback. In this case we might invoke the callback with a good response after invoking it with a timeout error. If we strictly want to avoid that, we can check here if the `state` is already `timer_expired` and if so, do nothing.
- in the `cancel` returned by the watch API, we should set the state to `cancelled`.

I feel this might be cleaner. What do you think?

File: xds/internal/client/client_callback.go (hunk 4 above)
Comment (menghanl) [reply xds/internal/client/client_callback.go]: This error is deleted.

File: xds/internal/client/client_callback.go (hunk 3 above)
Comment (menghanl) [reply xds/internal/client/client_callback.go]: Done. Also added methods so we remember to check the state before scheduling callback.
